pkg/admin/util/reflection/reflection_test.go

```diff
@@ -0,0 +1,520 @@
+
+	maxTry := 3
+	var err error
+	for i := 0; i < maxTry; i++ {
+		err = tripleRef.Dail(context.Background())
+		if err == nil {
+			break
+		}
+		time.Sleep(time.Second * 5)
+	}
+	if err != nil {
+		panic(err)
+	}
+}
+
+func TestListServices(t *testing.T) {
+	lists := []struct {
+		name     string
+		ref      RPCReflection
+		services []string
+	}{
+		{
+			name:     "grpc",
+			ref:      grpcRef,
+			services: []string{"greet.GreetService", "grpc.reflection.v1.ServerReflection", "grpc.reflection.v1alpha.ServerReflection"},
+		},
+		{
+			name:     "triple",
+			ref:      tripleRef,
+			services: []string{"greet.GreetService", "grpc.health.v1.Health", "grpc.reflection.v1alpha.ServerReflection"},
+		},
+	}
+
+	for _, list := range lists {
+		t.Run(list.name, func(t *testing.T) {
+			services, err := list.ref.ListServices()
+			if err != nil {
+				t.Error(err)
+				return
+			}
+
+			t.Logf("name:%#v services:%#v", list.name, services)
+			assert.Equal(t, services, list.services)
+		})
+	}
+}
+
+func TestListMethods(t *testing.T) {
+	lists := []struct {
+		name    string
+		ref     RPCReflection
+		service string
+		methods []string
+	}{
+		{
+			name:    "grpc",
+			ref:     grpcRef,
+			service: "greet.GreetService",
+			methods: []string{"greet.GreetService.Greet", "greet.GreetService.GreetClientStream", "greet.GreetService.GreetServerStream", "greet.GreetService.GreetStream"},
+		},
+		{
+			name:    "triple",
+			ref:     tripleRef,
+			service: "greet.GreetService",
+			methods: []string{"greet.GreetService.Greet", "greet.GreetService.GreetClientStream", "greet.GreetService.GreetServerStream", "greet.GreetService.GreetStream"},
+		},
+	}
+
+	for _, list := range lists {
+		t.Run(list.name, func(t *testing.T) {
+			methods, err := list.ref.ListMethods(list.service)
+			if err != nil {
+				t.Error(err)
+				return
+			}
+
+			t.Logf("name:%#v service:%#v methods:%#v", list.name, list.service, methods)
+			assert.Equal(t, methods, list.methods)
+		})
+	}
+}
+
+func TestMethodDescribeString(t *testing.T) {
+	lists := []struct {
+		name    string
+		ref     RPCReflection
+		service string
+		methods []string
+	}{
+		{
+			name:    "grpc",
+			ref:     grpcRef,
+			service: "greet.GreetService",
+			methods: []string{"greet.GreetService.Greet", "greet.GreetService.GreetClientStream", "greet.GreetService.GreetServerStream", "greet.GreetService.GreetStream"},
+		},
+		{
+			name:    "triple",
+			ref:     tripleRef,
+			service: "greet.GreetService",
+			methods: []string{"greet.GreetService.Greet", "greet.GreetService.GreetClientStream", "greet.GreetService.GreetServerStream", "greet.GreetService.GreetStream"},
+		},
+	}
+
+	for _, list := range lists {
+		t.Run(list.name, func(t *testing.T) {
+			for _, method := range list.methods {
+				txt, err := list.ref.DescribeString(method)
+				if err != nil {
+					t.Errorf("error in DescribeString for %v", method)
+					return
+				}
+				t.Logf("name:%#v method:%#v txt:\n%#v", list.name, method, txt)
```

Comment:
If you feel that verifying the results directly is tedious, you can determine if the results of grpc and triple are the same (assuming that grpc's results are correct).

Reply:
Get it! Thank you😊